sources/core/Xenko.Core.Serialization/Serialization/UrlReference.cs

```diff
@@ -0,0 +1,94 @@
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Globalization;
+using System.Text;
+using Xenko.Core.Annotations;
+using Xenko.Core.Assets;
+using Xenko.Core.Serialization.Contents;
+using Xenko.Core.Serialization.Serializers;
+
+namespace Xenko.Core.Serialization
+{
+    /// <summary>
+    /// Represents a Url to an asset.
+    /// </summary>
+    [DataContract("urlref", Inherited = true)]
+    [DataStyle(DataStyle.Compact)]
+    [ReferenceSerializer]
+    [DataSerializer(typeof(UrlReferenceDataSerializer))]
+    public class UrlReference
```

Comment:
I was wondering if it wouldn't be better as a `struct`?
Especially since we also have to check for `IsEmpty` so there is two ways to have invalid values currently.
Could remove lot of null check as well.
Not sure if all the converters and editor code will like it, but might be worth a try don't you think?

Reply:
`AttachedReferenceManager` requires `objects` because it uses a `ConditionalWeakTable` to attach references. Ideally the `UrlReference` types would not have a default constructor so you could not instantiate "empty" ones. But various parts of the asset stuff and UI require a default constructor for them to work.  It would require a lot more changes to accomplish this without default constructors.

I have created an interface and a base class so the `UrlReference` and `UrlReference<T>` types can both be `sealed`. I have also tidied up a bunch of things:
* Fix some formatting (remove blank likes etc.)
* Fixed/added doc comments
* Fixed some exceptions being thrown (Had parameters around the wrong way).
* Removed/sorted using statements
* Added License notice to the tops of files

Reply:
Good point about attached reference, makes sense. Thanks!

Wondering, could we do it the other way around and forbid to use null/empty string then?
i.e. if not set in editor, `UrlReference` instance itself is null. If instance is set, it has to have an actual valid `Url` value?

Note: if annoying/complex, no need to do that now.

Reply:
I did have a couple of solutions that didn't expose the default constructor that worked mostly. For example, make the default constructor `internal` and change how instantiation code works. I had a couple of different versions of this that "worked" but I wasn't too happy with the bits I changed. 

Though I might have just thought of an option....

But even still, doing it this way meant that using collection properties i.e. `List<UrlReference>` don't work because the UI code cannot instantiate an instance without a default constructor. So would either have to leave this feature out or change the UI code.

Reply:
That's true, let's keep it like this for now, thanks.